File: apps/agent-runtime/app/observability/sentry.py
"""Sentry opt-in para o agent-runtime.

No-op completo sem `SENTRY_DSN_AGENT_RUNTIME`: nenhuma conexão, nenhum overhead,
nenhuma exceção. Mesmo contrato dos apps Node (`@hm/api`/`@hm/workers`).
Idempotente — chamadas repetidas são ignoradas.

`init_sentry()` deve ser chamado uma vez no startup (lifespan), antes de servir
requests. O orchestrator faz o wire em `main.py`.
"""

from __future__ import annotations

import os

from app.logging import get_logger

logger = get_logger()

_initialized = False


def _sample_rate(env_key: str) -> float:
    """Lê uma taxa de amostragem [0,1] de env; default 0 (sem custo)."""
    raw = os.environ.get(env_key)
    if not raw:
        return 0.0
    try:
        value = float(raw)
    except ValueError:
        return 0.0
    return value if 0.0 <= value <= 1.0 else 0.0


def init_sentry() -> bool:
    """Inicializa o Sentry se houver DSN. Idempotente; devolve se está ativo."""
    global _initialized
    if _initialized:
        return True

    dsn = os.environ.get("SENTRY_DSN_AGENT_RUNTIME")
    if not dsn:
        return False

    import sentry_sdk

    sentry_sdk.init(
        dsn=dsn,
        environment=os.environ.get("NODE_ENV", "development"),
        release=os.environ.get("HM_RELEASE"),
        traces_sample_rate=_sample_rate("SENTRY_TRACES_SAMPLE_RATE"),
        # Nunca enviar PII por padrão (mensagens de contatos não vazam).
        send_default_pii=False,
    )
    _initialized = True
    logger.info("sentry: agent-runtime capture habilitado")
    return True


def is_sentry_enabled() -> bool:
    """True quando o Sentry foi efetivamente inicializado (DSN presente)."""
    return _initialized

File: apps/api/src/observability/sentry.ts
import * as Sentry from '@sentry/node';

/**
 * Sentry **opt-in** para a API. No-op completo sem `SENTRY_DSN_API`: nenhuma
 * conexão, nenhuma exceção, nenhum overhead. Mesmo padrão de `@hm/logger` OTel
 * (nada liga sem env). Idempotente — chamadas repetidas são ignoradas.
 *
 * O orchestrator chama `initSentry()` no topo do bootstrap (antes de criar o
 * app), e monta `sentryErrorHandler()` como ÚLTIMO middleware antes do error
 * handler central (em F10, o wire em app.ts é do orchestrator).
 */
let initialized = false;

export function initSentry(): boolean {
  if (initialized) return true;
  const dsn = process.env['SENTRY_DSN_API'];
  if (!dsn) return false;

  Sentry.init({
    dsn,
    environment: process.env['NODE_ENV'] ?? 'development',
    release: process.env['HM_RELEASE'],
    // Tracing opt-in: 0 por default (sem custo); ajustável por env.
    tracesSampleRate: sampleRate('SENTRY_TRACES_SAMPLE_RATE'),
    // Não enviar PII por padrão (telefones/e-mails de contatos não vazam).
    sendDefaultPii: false,
  });
  initialized = true;
  return true;
}

function sampleRate(envKey: string): number {
  const raw = process.env[envKey];
  if (!raw) return 0;
  const parsed = Number.parseFloat(raw);
  return Number.isFinite(parsed) && parsed >= 0 && parsed <= 1 ? parsed : 0;
}

/** True quando o Sentry foi efetivamente inicializado (DSN presente). */
export function isSentryEnabled(): boolean {
  return initialized;
}

/** Contexto opcional anexado ao evento (tags de correlação, ex.: `ref`, `workspaceId`). */
export interface CaptureContext {
  /** Tags indexáveis/pesquisáveis no Sentry. Valores não-string são ignorados. */
  readonly tags?: Readonly<Record<string, string | undefined>>;
}

/**
 * Captura uma exceção manualmente (no-op se desabilitado). Útil em catch-blocks
 * que tratam o erro mas ainda querem reportá-lo. As `tags` viram dimensões
 * pesquisáveis no Sentry sem contaminar o escopo global (usa `withScope`).
 *
 * Cuidado com PII: passe apenas identificadores de correlação (ref, workspaceId),
 * nunca telefone/e-mail/conteúdo de mensagem.
 */
export function captureException(error: unknown, context?: CaptureContext): void {
  if (!initialized) return;
  const tags = context?.tags;
  if (!tags) {
    Sentry.captureException(error);
    return;
  }
  Sentry.withScope((scope) => {
    for (const [key, value] of Object.entries(tags)) {
      if (typeof value === 'string' && value.length > 0) scope.setTag(key, value);
    }
    Sentry.captureException(error);
  });
}

/**
 * Error handler Express do Sentry (4-args). No-op seguro quando desabilitado —
 * apenas repassa ao próximo handler. Deve preceder o `errorHandler` central.
 */
export function sentryErrorHandler(): ReturnType<typeof Sentry.expressErrorHandler> {
  return Sentry.expressErrorHandler();
}

export { Sentry };
